### ISSUE TITLE
T2CharString.draw() does not emit final closePath() for CFF2 charstrings

### ISSUE THREAD

**Issue by justvanrossum** [fonttools/fonttools#1899]
This is likely caused by `endchar` being implicit in CFF2. Investigating.

**Comment by justvanrossum** [comment]
Adding this at the end of `SimpleT2Decompiler.execute()` fixes it:

```python
		if getattr(self.private, "_isCFF2", False):
			self.op_endchar(None)
```

But I'm not sure that's the cleanest solution.

**Comment by anthrotype** [comment]
maybe you could just call `pen.closePath()` from `_TTGlyphCFF.draw()` (or even better define a new `_TTGlyphCFF2` class in `fontTools.ttLib.ttFont` module and return that from `TTFont.getGlyphSet`).

**Comment by anthrotype** [comment]
@justvanrossum  how about a patch like this?

```diff
diff --git a/Lib/fontTools/misc/psCharStrings.py b/Lib/fontTools/misc/psCharStrings.py
index b894653d..4b6c83d8 100644
--- a/Lib/fontTools/misc/psCharStrings.py
+++ b/Lib/fontTools/misc/psCharStrings.py
@@ -508,6 +508,18 @@ class T2OutlineExtractor(T2WidthExtractor):
                self.currentPoint = (0, 0)
                self.sawMoveTo = 0

+       def execute(self, charString):
+               super().execute(charString)
+
+               # endchar was removed in CFF2 (it's now implied); ensure the last path is closed
+               try:
+                       isCFF2 = self.private._isCFF2
+               except AttributeError:
+                       pass
+               else:
+                       if isCFF2:
+                               self.endPath()
+
        def _nextPoint(self, point):
                x, y = self.currentPoint
                point = x + point[0], y + point[1]
```

I think it's the job of T2OutlineExtractor to call the pen's closePath, not SimpleT2Decompiler to call endchar operator which is not there.

I'm doing `try.. except AttributeError` instead of `getattr(self.private, "_isCFF2", False)` because even `self.private` may not be there.. For some arcane reasons the `T1OutlineExtractor` is a sub-class of `T2OutlineExtractor` but it has a different constructor and doesn't set self.private attribute...

Do you have a test case?

**Comment by anthrotype** [comment]
the `T2OutlineExtractor.closePath` method can be safely called more than once, as it's no-op when it has not seen a previous moveTo

**Comment by justvanrossum** [comment]
> how about a patch like this?

I like that, that seems the cleanest, most correct way.

> Do you have a test case?

No yet in a sharable way, I think we can build one based on `TestCFF2_ExtraLight.ttx` or similar.

I can work on a PR, but not right now (it's not a blocking issue for me).

> For some arcane reasons the T1OutlineExtractor is a sub-class of T2OutlineExtractor

That was me... I remember there was a practical reason for that, but of course I don't remember the details :(

**Comment by anthrotype** [comment]
this #1932 may be relevant here. @justvanrossum should I revert it?

**Comment by justvanrossum** [comment]
> this #1932 may be relevant here. @justvanrossum should I revert it?

Right, hm. That super call introduces some attributes that are not relevant to `T1CharString`. Although it would make your proposed fix above a little simpler: no try/except needed as "private" will exist even in `T1CharString`.

It's a bit of a mess either way, so I'm fine with leaving the `super()` in.

**Comment by skef** [comment]
I'm hitting this in (what passes for) the real world. Should I look through these notes and try to make a compromise PR or is that better left to others?

**Comment by justvanrossum** [comment]
Hmmm this seems a duplicate of #2455, which just got fixed.

**Comment by skef** [comment]
Ah, yeah, that is a recent enough change that I definitely haven't seen it yet. I'll pull it over and try it out.

**Comment by skef** [comment]
Hmm, well, that fix doesn't resolve the issue in my code because my object inherits from `T2OutlineExtractor` to do some extra work. I can add my own trailing `endPath()` call, of course.

However, `closePath()` support enters the object hierarchy at `T2WidthExtractor`. Wouldn't the following be a bit more self-contained? It passes the test suite:

```
diff --git a/Lib/fontTools/misc/psCharStrings.py b/Lib/fontTools/misc/psCharStrings.py
index 6014abc2..a42d7b5d 100644
--- a/Lib/fontTools/misc/psCharStrings.py
+++ b/Lib/fontTools/misc/psCharStrings.py
@@ -508,6 +508,10 @@ class T2OutlineExtractor(T2WidthExtractor):
                self.currentPoint = (0, 0)
                self.sawMoveTo = 0
 
+       def execute(self, charString):
+               super().execute(charString)
+               self.endPath()
+
        def _nextPoint(self, point):
                x, y = self.currentPoint
                point = x + point[0], y + point[1]
@@ -980,7 +984,6 @@ class T2CharString(object):
                                self.private.nominalWidthX, self.private.defaultWidthX,
                                self.private)
                extractor.execute(self)
-               extractor.endPath()
                self.width = extractor.width
 
        def calcBounds(self, glyphSet):
```

**Comment by skef** [comment]
Well, hmm, maybe that doesn't work ...

**Comment by skef** [comment]
Unsettling that that passed the test suite... I guess lots of extra closePaths would only be caught by a specific sort of test. 

Going down this road needs more change, along these lines:

```
diff --git a/Lib/fontTools/cffLib/__init__.py b/Lib/fontTools/cffLib/__init__.py
index fc82bb27..9eb328f3 100644
--- a/Lib/fontTools/cffLib/__init__.py
+++ b/Lib/fontTools/cffLib/__init__.py
@@ -50,7 +50,7 @@ class _DesubroutinizingT2Decompiler(psCharStrings.SimpleT2Decompiler):
                psCharStrings.SimpleT2Decompiler.__init__(self, localSubrs, globalSubrs,
                                                                               private)
 
-       def execute(self, charString):
+       def execute(self, charString, isSubr=False):
                self.need_hintcount = True  # until proven otherwise
                for op_name in self.stop_hintcount_ops:
                        setattr(self, op_name, self.stop_hint_count)
@@ -62,13 +62,14 @@ class _DesubroutinizingT2Decompiler(psCharStrings.SimpleT2Decompiler):
                        # counting hints pairs.
                        if self.need_hintcount and self.callingStack:
                                try:
-                                       psCharStrings.SimpleT2Decompiler.execute(self, charString)
+                                       psCharStrings.SimpleT2Decompiler.execute(self, charString,
+                                                                                isSubr)
                                except StopHintCountEvent:
                                        del self.callingStack[-1]
                        return
 
                charString._patches = []
-               psCharStrings.SimpleT2Decompiler.execute(self, charString)
+               psCharStrings.SimpleT2Decompiler.execute(self, charString, isSubr)
                desubroutinized = charString.program[:]
                for idx, expansion in reversed(charString._patches):
                        assert idx >= 2
diff --git a/Lib/fontTools/misc/psCharStrings.py b/Lib/fontTools/misc/psCharStrings.py
index 6014abc2..58408fa5 100644
--- a/Lib/fontTools/misc/psCharStrings.py
+++ b/Lib/fontTools/misc/psCharStrings.py
@@ -278,7 +278,7 @@ class SimpleT2Decompiler(object):
                self.hintMaskBytes = 0
                self.numRegions = 0
 
-       def execute(self, charString):
+       def execute(self, charString, isSubr=False):
                self.callingStack.append(charString)
                needsDecompilation = charString.needsDecompilation()
                if needsDecompilation:
@@ -335,12 +335,12 @@ class SimpleT2Decompiler(object):
        def op_callsubr(self, index):
                subrIndex = self.pop()
                subr = self.localSubrs[subrIndex+self.localBias]
-               self.execute(subr)
+               self.execute(subr, True)
 
        def op_callgsubr(self, index):
                subrIndex = self.pop()
                subr = self.globalSubrs[subrIndex+self.globalBias]
-               self.execute(subr)
+               self.execute(subr, True)
 
        def op_hstem(self, index):
                self.countHints()
@@ -508,6 +508,11 @@ class T2OutlineExtractor(T2WidthExtractor):
                self.currentPoint = (0, 0)
                self.sawMoveTo = 0
 
+       def execute(self, charString, isSubr=False):
+               super().execute(charString, isSubr)
+               if not isSubr:
+                       self.endPath()
+
        def _nextPoint(self, point):
                x, y = self.currentPoint
                point = x + point[0], y + point[1]
@@ -858,7 +863,7 @@ class T1OutlineExtractor(T2OutlineExtractor):
        def op_callsubr(self, index):
                subrIndex = self.pop()
                subr = self.subrs[subrIndex]
-               self.execute(subr)
+               self.execute(subr, True)
        def op_callothersubr(self, index):
                subrIndex = self.pop()
                nArgs = self.pop()
@@ -980,7 +985,6 @@ class T2CharString(object):
                                self.private.nominalWidthX, self.private.defaultWidthX,
                                self.private)
                extractor.execute(self)
-               extractor.endPath()
                self.width = extractor.width
 
        def calcBounds(self, glyphSet):
diff --git a/Lib/fontTools/subset/cff.py b/Lib/fontTools/subset/cff.py
index 0dcb7975..9eedb22b 100644
--- a/Lib/fontTools/subset/cff.py
+++ b/Lib/fontTools/subset/cff.py
@@ -240,11 +240,11 @@ class _DehintingT2Decompiler(psCharStrings.T2WidthExtractor):
                        self, localSubrs, globalSubrs, nominalWidthX, defaultWidthX)
                self.private = private
 
-       def execute(self, charString):
+       def execute(self, charString, isSubr=False):
                old_hints = charString._hints if hasattr(charString, '_hints') else None
                charString._hints = self.Hints()
 
-               psCharStrings.T2WidthExtractor.execute(self, charString)
+               psCharStrings.T2WidthExtractor.execute(self, charString, isSubr)
 
                hints = charString._hints
 

```
This works in the sense that it passes the test suite as well as the tests for my derived object. However, it changes an interface and may not be what you prefer.

**Comment by justvanrossum** [comment]
I'm curious why you would need to subclass `T2OutlineExtractor` in the first place.

As you have seen, the inheritance hierarchy is a little strange here at best (it's pretty old code), and I think client code shouldn't ever need to subclass further.

(Also: the general argument that composition is better inheritance. Which, in my 2022 opinion, is almost always true.)

We could even argue that `T2OutlineExtractor` is an internal class, and subclassing it is not supported.

**Comment by skef** [comment]
It seems like in this case any sub-classing is beside the point. Let's say that I were to reorganize the code (which is based on other code that has been inheriting from `T2OutlineExtractor` for years, but could be changed). Even then, my code and any other code that *uses* the outline extractor outside of a `T2CharString` has to add this (currently undocumented) extra call to `endPath()` after calling `execute()` or it will be broken for CFF2. That includes any code using the object in a composition pattern. So it seems worth fixing regardless.

And, of course, if you feel that subclassing is the less desirable approach that implies that any subclassing is more or less the problem of the subclasser rather than fontTools. If that's true then there's less reason to be wary of my patch, given that the `isSubr` parameter has a default. (I suppose delegators calling `execute()` directly will still be screwed and have to update their code to avoid spurious path-closings. But that only happens in the current hierarchy at low-levels anyway. ) 

If you're curious the "based on" code is [here](https://github.com/adobe-type-tools/psautohint/blob/b20dea4447b2bd48aeb85aa6e898c80f4842d425/python/psautohint/otfFont.py#L69). The adapted code is a bit simpler.

**Comment by justvanrossum** [comment]
Thanks. I'm afraid this mostly gives me a headache at the moment, and I'd gladly defer judgement to others if possible. I honestly don't see an obvious path forward here.

(Headache co-triggered by my own code design decisions made more than 20 years ago...)

**Comment by skef** [comment]
In that case I guess I'll turn my patch into a PR so that interested parties can discuss the questions there. (With full understanding that nothing may come of it.)

**Comment by behdad** [comment]
> Thanks. I'm afraid this mostly gives me a headache at the moment, and I'd gladly defer judgement to others if possible. I honestly don't see an obvious path forward here.

I feel the same unfortunately.

**Comment by justvanrossum** [comment]
Fixed in #2580. Thank you, @skef! Much appreciated.
